numba/core/typing/new_builtins.py

```diff
@@ -127,53 +131,506 @@ def generic(self, args, kws):
+    if argtys in op_cache[op]:
+        return op_cache[op][argtys]
+
+    argvals = [argty.cast_python_value(0) for argty in argtys]
+    res_val = getattr(argvals[0], op)(*argvals[1:])
+    if res_val is not NotImplemented:
+        res_val = FROM_DTYPE[res_val.dtype]
+    op_cache[op][argtys] = res_val
+    return res_val
+
+
+@overload_method(types.NumPyBoolean, "__add__")
+@overload_method(types.NumPyInteger, "__add__")
+@overload_method(types.NumPyFloat, "__add__")
+@overload_method(types.NumPyComplex, "__add__")
+def np__add__(self, other):
+    final_ty = find_np_res_type("__add__", binop_cache, (self, other))
+
+    if final_ty is NotImplemented:
+        def impl(self, other):
+            return NotImplemented
+        return impl
+
+    if 'bool' in final_ty.name:
+        add_func = np_bool_add_bool
+    elif 'int' in final_ty.name:
+        add_func = int_add_int
+    elif 'float' in final_ty.name:
+        add_func = float_add_float
+    elif 'complex' in final_ty.name:
+        add_func = complex_add_complex
+    else:
+        raise TypeError(f"Addition function not defined for {self} and {other}")
+
+    # evaluate type
+    def impl(self, other):
+        return add_func(final_ty(self), final_ty(other))
+    return impl
+
+
+@overload_method(types.NumPyBoolean, "__radd__")
+@overload_method(types.NumPyInteger, "__radd__")
+@overload_method(types.NumPyFloat, "__radd__")
+@overload_method(types.NumPyComplex, "__radd__")
+def np__radd__(self, other):
+    final_ty = find_np_res_type("__radd__", binop_cache, (self, other))
+
+    if final_ty is NotImplemented:
+        def impl(self, other):
+            return NotImplemented
+        return impl
+
+    if 'bool' in final_ty.name:
+        add_func = np_bool_add_bool
+    elif 'int' in final_ty.name:
+        add_func = int_add_int
+    elif 'float' in final_ty.name:
+        add_func = float_add_float
+    elif 'complex' in final_ty.name:
+        add_func = complex_add_complex
+    else:
+        raise TypeError(f"Addition function not defined for {self} and {other}")
+
+    # evaluate type
+    def impl(self, other):
+        return add_func(final_ty(self), final_ty(other))
+    return impl
+
+
+def generate_binop(op_func, slot, rslot, opchar):
+
+    # based loosely on:
+    # https://github.com/python/cpython/blob/1a1e013a4a526546c373afd887f2e25eecc984ad/Objects/abstract.c#L916-L977
+    def binary_op(v, w):
+        pass
+
+    @overload(binary_op)
+    def ol_binary_op_number(v, w):
+        if not isinstance(v, (types.Number, types.Boolean)) or not isinstance(w, (types.Number, types.Boolean)):
+            return
+
+        # this really needs to only deal with number and sequence types, can
+        # either make unsupported typse a compilation error, or fall back to
+        # making it a runtime error like "unsupported operand for types..."
+        is_subclass = issubclass(
+            type(w.cast_python_value(0)),
+            type(v.cast_python_value(0))
+        )
```

Comment:
Need to handle types that don't have `.cast_python_value`. Perhaps, add a `.issubclass()` method to numba Type class with a default impl that uses the class hierarchy of the Types.

My hack to make it work:
```diff
@@ -584,4 +589,7 @@ def generate_binop(op_func, slot, rslot, opchar):
-        is_subclass = issubclass(
-            type(w.cast_python_value(0)),
-            type(v.cast_python_value(0))
-        )
+        try:
+            is_subclass = issubclass(
+                type(w.cast_python_value(0)),
+                type(v.cast_python_value(0))
+            )
+        except NotImplementedError:
+            is_subclass = False
```